nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# The following is largely based on code from https://github.com/lucidrains/stylegan2-pytorch
+
+import math
+from random import random, randrange
+from typing import Dict, List, Optional, TypedDict, Union
+
+import torch
+import torch.nn.functional as F
+import torchvision
+from einops import rearrange, repeat
+from omegaconf import DictConfig, OmegaConf
+from pytorch_lightning import Trainer
+from pytorch_lightning.loggers import TensorBoardLogger, WandbLogger
+from torch import Tensor, nn
+from torch.autograd import grad as torch_grad
+from torch.utils.tensorboard.writer import SummaryWriter
+
+import nemo
+from nemo.collections.tts.models import FastPitchModel
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.collections.tts.modules.spectrogram_enhancer import Discriminator, Generator, mask
+from nemo.core import Exportable, ModelPT
+from nemo.utils import logging
+
+
+def type_as_recursive(e, source: Tensor):
+    if isinstance(e, (list, tuple)):
+        return [type_as_recursive(elem, source) for elem in e]
+    elif isinstance(e, dict):
+        return {key: type_as_recursive(value, source) for key, value in e.items()}
+    elif isinstance(e, Tensor):
+        return e.type_as(source)
+    else:
+        return e
+
+
+def gradient_penalty(images, output, weight=10):
+    batch_size = images.shape[0]
+    gradients = torch_grad(
+        outputs=output,
+        inputs=images,
+        grad_outputs=torch.ones(output.size(), device=images.device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    gradients = gradients.reshape(batch_size, -1)
+    return weight * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
+
+
+def calc_pl_lengths(styles, images):
+    device = images.device
+    num_pixels = images.shape[2] * images.shape[3]
+    pl_noise = torch.randn(images.shape, device=device) / math.sqrt(num_pixels)
+    outputs = (images * pl_noise).sum()
+
+    pl_grads = torch_grad(
+        outputs=outputs,
+        inputs=styles,
+        grad_outputs=torch.ones(outputs.shape, device=device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    return (pl_grads ** 2).sum(dim=2).mean(dim=1).sqrt()
+
+
+def gen_hinge_loss(fake):
```

Comment:
name of the method and what it is doing are not the same. can be confusing. You can get rid of the method and replace with just `variable.mean()`